File: apps/worker-notification/channels/email.ts
import type { NotificationChannel, NotificationPayload } from "./types";
import FormData from "form-data"; 
import Mailgun from "mailgun.js"; 
import { env } from "../env";

const MAILGUN_API_KEY = env.MAILGUN_API_KEY;


export class EmailChannel implements NotificationChannel {
    async send(payload: NotificationPayload): Promise<void> {
        console.log(`Website ${payload.websiteId} is ${payload.eventType} at ${payload.occurredAt}`);
        await sendEmailViaMailgun(payload);
    }
}

async function sendEmailViaMailgun( payload : NotificationPayload ) {
  const mailgun = new Mailgun(FormData);
  const mg = mailgun.client({
    username: "api",
    key: MAILGUN_API_KEY,
  });
  try {
    const data = await mg.messages.create("sandboxf0ae1aa71b334d07abbef5327293e589.mailgun.org", {
      from: "Mailgun Sandbox <postmaster@sandboxf0ae1aa71b334d07abbef5327293e589.mailgun.org>",
      to: [payload.email],
      subject: `Website Alert ${payload.eventType}`,
      text: `Your website ${payload.websiteId} is ${payload.eventType} at ${payload.occurredAt}`,
    });

    console.log("Email sent",data); // logs response data
  } catch (error) {
    console.error(`Mailgun failed`,error); //logs any error
  }
}

File: apps/api-go/internal/channels/email.go
package channels

import (
	"context"
	"fmt"

	"github.com/mailgun/mailgun-go/v5"
)

type EmailChannel struct {
	mg     mailgun.Mailgun
	domain string
	sender string
}

func NewEmailChannel(
	apiKey string,
	domain string,
	sender string,
) *EmailChannel {
	mg := mailgun.NewMailgun(apiKey)

	return &EmailChannel{
		mg:     mg,
		domain: domain,
		sender: sender,
	}
}

func (ec *EmailChannel) Send(
	ctx context.Context,
	payload NotificationPayload,
) (string, error) {
	subject := fmt.Sprintf("Website Alert %s", payload.EventType)

	body := fmt.Sprintf(
		"Your website %s is %s at %s",
		payload.WebsiteID,
		payload.EventType,
		payload.OccurredAt,
	)

	msg := mailgun.NewMessage(
		ec.sender,
		subject,
		body,
		payload.Target, // email address
	)

	resp, err := ec.mg.Send(ctx, msg)
	if err != nil {
		return "", err
	}
	return resp.ID, nil
}
